lang/cs/Org.Apache.REEF.Network/Elastic/Topology/Physical/Default/OneToNTopology.cs

```diff
@@ -0,0 +1,275 @@
+using System.Collections.Concurrent;
+using System.Linq;
+using Org.Apache.REEF.Utilities.Attributes;
+
+namespace Org.Apache.REEF.Network.Elastic.Topology.Physical.Default
+{
+    /// <summary>
+    /// Base class for topologies following a one to N communication pattern.
+    /// </summary>
+    [Unstable("0.16", "API may change")]
+    internal abstract class OneToNTopology : OperatorTopologyWithDefaultCommunication
+    {
+        protected static readonly Logger Log = Logger.GetLogger(typeof(OneToNTopology));
+
+        protected readonly ConcurrentDictionary<string, byte> _nodesToRemove =
+            new ConcurrentDictionary<string, byte>();
+
+        protected readonly ManualResetEvent _topologyUpdateReceived;
+        protected readonly bool _piggybackTopologyUpdates;
+
+        /// <summary>
+        /// Construct a one to N topology.
+        /// </summary>
+        /// <param name="stageName">The stage name the topology is working on</param>
+        /// <param name="taskId">The identifier of the task the topology is running on</param>
+        /// <param name="rootTaskId">The identifier of the root note in the topology</param>
+        /// <param name="operatorId">The identifier of the operator for this topology</param>
+        /// <param name="children">The list of nodes this task has to send messages to</param>
+        /// <param name="piggyback">Whether to piggyback topology update messages to data message</param>
+        /// <param name="retry">How many times the topology will retry to send a message</param>
+        /// <param name="timeout">After how long the topology waits for an event</param>
+        /// <param name="disposeTimeout">Maximum wait time for topology disposal</param>
+        /// <param name="commLayer">Layer responsible for communication</param>
+        /// <param name="checkpointLayer">Layer responsible for saving and retrieving checkpoints</param>
+        public OneToNTopology(
+            string stageName,
+            string taskId,
+            string rootTaskId,
+            int operatorId,
+            ISet<int> children,
+            bool piggyback,
+            int retry,
+            int timeout,
+            int disposeTimeout,
+            DefaultCommunicationLayer commLayer) : base(
+                stageName,
+                taskId,
+                rootTaskId,
+                operatorId,
+                commLayer,
+                retry,
+                timeout,
+                disposeTimeout)
+        {
+            _topologyUpdateReceived = new ManualResetEvent(RootTaskId != taskId);
+
+            _commLayer.RegisterOperatorTopologyForTask(this);
+            _commLayer.RegisterOperatorTopologyForDriver(this);
+
+            _piggybackTopologyUpdates = piggyback;
+
+            foreach (var child in children)
+            {
+                var childTaskId = Utils.BuildTaskId(StageName, child);
+
+                _children.TryAdd(child, childTaskId);
```

Comment:
in fact, we can pass children into the `OperatorTopologyWithDefaultCommunication` constructor and initialize the dictionary once:
```C#
protected OperatorTopologyWithDefaultCommunication(
    // ...
    int disposeTimeout,
    IEnumerable<KeyValuePair<int, string>> children = null)
{
    _children = children == null
                ? new ConcurrentDictionary<int, string>()
                : new ConcurrentDictionary<int, string>(children);
```
and here do something like
```C#
: base(
    // ...
    disposeTimeout,
    children.Select(child =>
        new KeyValuePair<int, string>(child, Utils.BuildTaskId(stageName, child))))
```

Reply:
I am not sure about this. I want to maintain the population of `_children` in the subclass because it is operator specific.